Allow to apply abritrary affine transforms

dirkschulze opened the pull request:
Aa alternative to (a otherwise useless) `WD_MATRIX`  it would be possible to pass the 6 floats as arguments to `wdSetWorldTransform`. However, the current implementation is more like the actual GDI+ API.

mity commented on src/backend-gdix.c:
```diff
@@ -113,6 +113,8 @@ gdix_init(void)
     GPA(SetSmoothingMode, (dummy_GpGraphics*, dummy_GpSmoothingMode));
     GPA(TranslateWorldTransform, (dummy_GpGraphics*, float, float, dummy_GpMatrixOrder));
+    GPA(SetWorldTransform, (dummy_GpGraphics*, dummy_GpMatrix*, dummy_GpMatrixOrder));
+    GPA(CreateMatrix2, (float, float, float, float, float, float, dummy_GpMatrix**));
```
I guess that if you need `GdipCreateMatrix2()`, you should likely also need `GdipDeleteMatrix()`.

dirkschulze replied:
Forgot that the initialized matrix needs to get deleted.

mity commented on src/canvas.c:
```diff
@@ -466,6 +466,31 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
+        gdix_canvas_t* c = (gdix_canvas_t*) hCanvas;
+
+        dummy_GpMatrix* matrix;
+        gdix_vtable->fn_CreateMatrix2(pMatrix->m11, pMatrix->m12,
```
Seems as the GDI+ matrix object is leaking here.

dirkschulze replied:
indeed

mity commented on src/canvas.c:
```diff
@@ -466,6 +466,31 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
+        gdix_vtable->fn_CreateMatrix2(pMatrix->m11, pMatrix->m12,
+                                      pMatrix->m21, pMatrix->m22,
+                                      pMatrix->dx, pMatrix->dy, &matrix);
+        gdix_vtable->fn_SetWorldTransform(c->graphics, matrix, dummy_MatrixOrderAppend);
```
I am confused. Do you want `wdSetWorldTransform()` to 
1. just apply the provided transformation matrix, or 
2. should it combine it with the already set transformation matrix?

Both can make good sense, and I can imagine to have functions for both. But now it is some strange hybrid and it even makes something different for each back-end:
* The function name `wdSetWorldTransform()` imho indicates more the behavior 1.
* For D2D back-end, `d2d_apply_transform()` implements behavior 2.
* Fir GDI+ back-end, `GdipSetWorldTransform()` implements behavior 1, but your attempt to use the flag `dummy_MatrixOrderAppend` indicates that maybe you are attempting to implement behavior 2. If you want behavior 2, you should likely use `GdipMultiplyWorldTransform()` instead. 
   
Also note WinDrawLib also sometimes applies some transformation internally:
   * For D2D backend, there is the `0.5f, 0.5f` translation so both back-ends shared the coordinate system, `[0,0]` meaning left-top pixel.
   * With a RTL flag, the canvas automatically applies left-to-right mirroring transformation.

If `wdSetWorldTransform()` is supposed to implement behavior 1, it should still imho still keep this internal initial transformations.

dirkschulze replied:
Ok, looking more into it it makes more sense to have a set instead of a multiply. This is because the interface is not a state machine as other graphic libraries use (with save/restore). However, the initial transform for rtl indeed should be kept.

mity commented on src/canvas.c:
```diff
@@ -466,6 +466,31 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
+        gdix_canvas_t* c = (gdix_canvas_t*) hCanvas;
+
+        dummy_GpMatrix* matrix;
+        gdix_vtable->fn_CreateMatrix2(pMatrix->m11, pMatrix->m12,
```
1. You are calling `GdipCreateMatrix2()` but the matrix objects seems to leak. You also need to call `GdipDeleteMatrix()` when you are done with it.

2. Most (if not all) WinDrawLib painting functions do not return anything (`void` return type). Because in my eyes, apps have no reasonable error handling for such functions but just still paint at least something. But we should still never cause correctly written app to e.g. crash, even if some GDI+ or Direct2D function fails. Please add some test whether the  `GdipCreateMatrix2()` succeeds. Passing `NULL` or invalid pointer into the other function below simply can not be allowed.

    In the case of matrix creation failure, just some debug message with `WD_TRACE()` and early `return` would be enough. Look to other code which creates any GDI+ object.

mity commented:
> Aa alternative to (a otherwise useless) `WD_MATRIX` it would be possible to pass the 6 floats as arguments to wdSetWorldTransform.

Generally I prefer functions which do not require to initialize many helper structures to pass many params. But imho `WD_MATRIX` is fine here. Nobody would remember the order of the 6 parameters, so using the matrix and its named members leads to more readable code.

mity commented on src/backend-gdix.h:
```diff
@@ -74,6 +74,8 @@ struct gdix_vtable_tag {
     int (WINAPI* fn_SetPixelOffsetMode)(dummy_GpGraphics*, dummy_GpPixelOffsetMode);
     int (WINAPI* fn_SetSmoothingMode)(dummy_GpGraphics*, dummy_GpSmoothingMode);
     int (WINAPI* fn_TranslateWorldTransform)(dummy_GpGraphics*, float, float, dummy_GpMatrixOrder);
+    int (WINAPI* fn_SetWorldTransform)(dummy_GpGraphics*, dummy_GpMatrix*, dummy_GpMatrixOrder);
```
`GdipSetWorldTransform()` takes only two params. Consult `<gdiplusflat.h>`. Maybe you want `GdipMultiplyWorldTransform()` instead?

dirkschulze replied:
Initially the argument was just too much. With the requirement for the initial rtl transformation it must be the mentioned matrix now.

mity commented on examples/draw-simple.c:
```diff
@@ -115,7 +126,7 @@ _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPTSTR lpCmdLine, int nC
     MSG msg;
 
-    wdPreInitialize(NULL, NULL, 0);
+    wdPreInitialize(NULL, NULL, WD_DISABLE_D2D);
```
A leftover from some testing? I wouldn't propagate `WD_DISABLE_D2D` in the examples. 

(If we want to allow it, it would be better to adapt the examples to accept some commandline option or env. var so both back-ends can be tested without recompilation, but that is off-topic for this PR.)

dirkschulze replied:
The changes here were not supposed to be committed :/

mity commented on src/canvas.c:
```diff
@@ -469,8 +469,14 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
 void
 wdSetWorldTransform(WD_HCANVAS hCanvas, const WD_MATRIX* pMatrix)
 {
+    if(pMatrix == NULL) {
```
Ouch. You've now made the code so super-careful that I am wondering whether I mislead you. Sorry if I did.

I just meant to just follow some code pattern like

```
res = alloc_resource();
if(alloc_resource() failed) {
   log_something();
   return;
}
use_resource(res);
free_resource(res);
```

My aim was just to make the guarantee that as long as any app uses the API correctly, nothing bad as memory corruption happens. It previously could if the GDI+ matrix creation failed. 

But feel free to keep it. It doesn't make any harm.

mity commented on examples/draw-simple.c:
```diff
@@ -27,11 +27,22 @@ MainWinPaintToCanvas(WD_HCANVAS hCanvas)
+        m.dx = 0;
+        m.dy = 0;
+        wdSetWorldTransform(hCanvas, &m);
+
```
Please no.

I wouldn't add this into "draw-simple.c" at all. It doesn't really show much when someone just runs the example without looking into the source. Users wouldn't likely even notice it presents anything with transformations. 

I am also afraid that "draw-simple.c" would gradually change into "draw-everything.c". "draw-simple.c" is typically the very 1st piece of code newcomers are studying when looking at the API and how to use it. Lets keep this example really simple. It should do only very basic thing besides minimal initialization and plumbing it to Win32 app code.

If you want this PR to add some example code (which would be nice but it's not required), it would be much much better to make a new example with is dedicated to transformations. It could cor example paint something like https://code-industry.net/wp-content/uploads/2016/05/matrix14.png, but it could be much simpler (e.g. just s square, one normal, and other ones with various transformations in effect, so it could present also the other transformation functions like e.g. `wdRotateWorld()` et al.

dirkschulze commented:
With the need to call `wdResetWorld` and then multiply the CTM with the passed transformation matrix, having a `wdSetWorldTransform` makes no real sense anymore.

I replaced `wdSetWorldTransform` with `wdMultiplyWorldTransform`. An author could always call `wdResetWorld` him/herself to get the same result. Therefore, `wdMultiplyWorldTransform` covers `wdSetWorldTransform` and is more powerful.

I also added a new test. The result looks like this https://codepen.io/krit/pen/QPBWRz (cut off in Edge but works everywhere else.)

dirkschulze commented on src/canvas.c:
```diff
@@ -466,6 +466,47 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
 }
 
+void
+wdMultiplyWorldTransform(WD_HCANVAS hCanvas, const WD_MATRIX* pMatrix)
```
Tested with D2D and GDI. Added example looks the same on both.

mity commented:
> With the need to call wdResetWorld and then multiply the CTM with the passed transformation matrix,  having a wdSetWorldTransform makes no real sense anymore.

Good point.
 
> I replaced wdSetWorldTransform with wdMultiplyWorldTransform. An author could always call wdResetWorld him/herself to get the same result. Therefore, wdMultiplyWorldTransform covers wdSetWorldTransform and is more powerful.

Could you consider renaming `wdMultiplyWorldTransform()` to something what would be more consistent with the other transformation function names? Maybe `wdTransformWorld()`.

> I also added a new test. The result looks like this https://codepen.io/krit/pen/QPBWRz (cut off in Edge but works everywhere else.)

Looks good.

mity commented:
Merged. Thanks for the work.
